File: pegasusX/packages/mobile-android-kit/src/main/java/com/pegasusx/mobilekit/offline/OfflineHttpSemantics.kt
package com.pegasusx.mobilekit.offline

/**
 * Shared HTTP flush semantics for durable offline queues (§8.8).
 * 2xx / 409 → ACK (purge); 408/429/5xx → retry; other 4xx → DEAD/discard.
 */
object OfflineHttpSemantics {
    const val STATUS_PENDING = "PENDING"
    const val STATUS_DEAD = "DEAD"
    const val MAX_ATTEMPTS_DEFAULT = 8

    fun normalizeEndpoint(endpoint: String): String =
        endpoint.trim().removePrefix("/").removePrefix("api/")

    fun isRetryableHttp(code: Int): Boolean =
        code == 408 || code == 429 || code in 500..599

    fun isSuccessHttp(code: Int): Boolean =
        code in 200..299 || code == 409

    fun isDeadHttp(code: Int): Boolean =
        code in 400..499 && !isRetryableHttp(code) && code != 409

    fun isNetworkEnqueueable(error: Throwable): Boolean = when (error) {
        is java.io.IOException -> true
        else -> {
            val msg = error.message.orEmpty().lowercase()
            "unable to resolve host" in msg ||
                "failed to connect" in msg ||
                "timeout" in msg ||
                "connection reset" in msg
        }
    }

    enum class FlushOutcome {
        ACK,
        RETRY,
        DEAD,
    }

    fun outcomeForHttp(code: Int): FlushOutcome = when {
        isSuccessHttp(code) -> FlushOutcome.ACK
        isRetryableHttp(code) -> FlushOutcome.RETRY
        else -> FlushOutcome.DEAD
    }
}

File: pegasusX/packages/mobile-ios-kit/Sources/PegasusKit/Offline/OfflineHttpSemantics.swift
import Foundation

/// Shared HTTP flush semantics for durable offline queues (§8.8).
public enum OfflineHttpSemantics {
    public static let statusPending = "PENDING"
    public static let statusDead = "DEAD"
    public static let maxAttemptsDefault = 8

    public static func normalizeEndpoint(_ endpoint: String) -> String {
        var ep = endpoint.trimmingCharacters(in: .whitespacesAndNewlines)
        if ep.hasPrefix("/") { ep.removeFirst() }
        if ep.hasPrefix("api/") { ep = String(ep.dropFirst(4)) }
        return ep
    }

    public static func isRetryableHTTP(_ code: Int) -> Bool {
        code == 408 || code == 429 || (500...599).contains(code)
    }

    public static func isSuccessHTTP(_ code: Int) -> Bool {
        (200...299).contains(code) || code == 409
    }

    public static func isDeadHTTP(_ code: Int) -> Bool {
        (400...499).contains(code) && !isRetryableHTTP(code) && code != 409
    }

    public enum FlushOutcome: Sendable {
        case ack
        case retry
        case dead
    }

    public static func outcome(forHTTP code: Int) -> FlushOutcome {
        if isSuccessHTTP(code) { return .ack }
        if isRetryableHTTP(code) { return .retry }
        return .dead
    }
}
